[jit] fix str -> Device implicit conversions

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43213 [jit] fix str -> Device implicit conversions**

A reversed isSubtypeOf caused erroreous conversions to be inserted.

Differential Revision: [D23192787](https://our.internmc.facebook.com/intern/diff/D23192787)

zdevito committed:
[jit] fix str -> Device implicit conversions

A reversed isSubtypeOf caused erroreous conversions to be inserted.

[ghstack-poisoned]

eellison reviewed (approved):
Lgtm

zdevito committed:
Update on "[jit] fix str -> Device implicit conversions"


A reversed isSubtypeOf caused erroreous conversions to be inserted.

Differential Revision: [D23192787](https://our.internmc.facebook.com/intern/diff/D23192787)

[ghstack-poisoned]

zdevito committed:
Update on "[jit] fix str -> Device implicit conversions"


A reversed isSubtypeOf caused erroreous conversions to be inserted.

Differential Revision: [D23192787](https://our.internmc.facebook.com/intern/diff/D23192787)

[ghstack-poisoned]

eellison reviewed (approved):
Changes LGTM.

I don't think we need these calls to unwrapOptional anymore: https://github.com/pytorch/pytorch/blob/de1d108fe8760c60952ce9f2059ada91cc3bf0b0/torch/csrc/jit/frontend/schema_matching.cpp#L70

zdevito committed:
Update on "[jit] fix str -> Device implicit conversions"


A reversed isSubtypeOf caused erroreous conversions to be inserted.

Differential Revision: [D23192787](https://our.internmc.facebook.com/intern/diff/D23192787)

[ghstack-poisoned]
